File: apps/desktop/src/lib/aws-inventory.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Ali Shaikh

import type { WorkspaceSnapshot } from "@/types/backend";

export type AwsInventoryScope =
  | "s3"
  | "ec2"
  | "lambda"
  | "dynamodb"
  | "sqs"
  | "sns"
  | "rds"
  | "ecs"
  | "eks"
  | "cloudformation"
  | "eventbridge"
  | "route53"
  | "elb"
  | "apigateway"
  | "secrets"
  | "logs"
  | "iam";

const TAB_SCOPE_MAP: Record<string, AwsInventoryScope | undefined> = {
  s3: "s3",
  ec2: "ec2",
  lambda: "lambda",
  dynamodb: "dynamodb",
  sqs: "sqs",
  sns: "sns",
  rds: "rds",
  ecs: "ecs",
  eks: "eks",
  cloudformation: "cloudformation",
  eventbridge: "eventbridge",
  route53: "route53",
  elb: "elb",
  apigateway: "apigateway",
  secrets: "secrets",
  logs: "logs",
  iam: "iam",
};

export function awsInventoryScopeForTab(tabId: string): AwsInventoryScope | undefined {
  return TAB_SCOPE_MAP[tabId];
}

export function awsInventoryStatusMessage(
  workspace: WorkspaceSnapshot,
  scope: AwsInventoryScope,
): string | undefined {
  switch (scope) {
    case "s3":
      return workspace.s3StatusMessage;
    case "ec2":
      return workspace.ec2StatusMessage;
    case "lambda":
      return workspace.lambdaStatusMessage;
    case "dynamodb":
      return workspace.dynamodbStatusMessage;
    case "sqs":
      return workspace.sqsStatusMessage;
    case "sns":
      return workspace.snsStatusMessage;
    case "rds":
      return workspace.rdsStatusMessage;
    case "ecs":
      return workspace.ecsStatusMessage;
    case "eks":
      return workspace.eksStatusMessage;
    case "cloudformation":
      return workspace.cloudFormationStatusMessage;
    case "eventbridge":
      return workspace.eventBridgeStatusMessage;
    case "route53":
      return workspace.route53StatusMessage;
    case "elb":
      return workspace.elbStatusMessage;
    case "apigateway":
      return workspace.apiGatewayStatusMessage;
    case "secrets":
      return workspace.secretsManagerStatusMessage;
    case "logs":
      return workspace.logsStatusMessage;
    case "iam":
      return workspace.iamStatusMessage;
    default:
      return undefined;
  }
}

const DEFAULT_INVENTORY_LOADING_LABELS: Record<AwsInventoryScope, string> = {
  s3: "Loading S3 buckets...",
  ec2: "Loading EC2 regions...",
  lambda: "Loading Lambda functions...",
  dynamodb: "Loading DynamoDB tables...",
  sqs: "Loading SQS queues...",
  sns: "Loading SNS topics...",
  rds: "Loading RDS instances...",
  ecs: "Loading ECS clusters...",
  eks: "Loading EKS clusters...",
  cloudformation: "Loading CloudFormation stacks...",
  eventbridge: "Loading EventBridge buses...",
  route53: "Loading Route 53 hosted zones...",
  elb: "Loading load balancers...",
  apigateway: "Loading API Gateway APIs...",
  secrets: "Loading Secrets Manager secrets...",
  logs: "Loading CloudWatch log groups...",
  iam: "Loading IAM roles and policies...",
};

export function awsInventoryLoadingLabel(
  workspace: WorkspaceSnapshot,
  scope: AwsInventoryScope,
): string {
  const status = awsInventoryStatusMessage(workspace, scope)?.trim();
  if (status) {
    return status;
  }
  return DEFAULT_INVENTORY_LOADING_LABELS[scope];
}

export function awsInventoryLoaded(
  workspace: WorkspaceSnapshot,
  scope: AwsInventoryScope,
): boolean {
  switch (scope) {
    case "s3":
      return (workspace.s3Buckets?.length ?? 0) > 0 ||
        (workspace.s3StatusMessage ?? "").length > 0;
    case "ec2":
      return (workspace.ec2Regions?.length ?? 0) > 0 ||
        (workspace.ec2StatusMessage ?? "").length > 0;
    case "lambda":
      return (workspace.lambdaRegions?.length ?? 0) > 0 ||
        (workspace.lambdaStatusMessage ?? "").length > 0;
    case "dynamodb":
      return (workspace.dynamodbRegions?.length ?? 0) > 0 ||
        (workspace.dynamodbStatusMessage ?? "").length > 0;
    case "sqs":
      return (workspace.sqsRegions?.length ?? 0) > 0 ||
        (workspace.sqsStatusMessage ?? "").length > 0;
    case "sns":
      return (workspace.snsRegions?.length ?? 0) > 0 ||
        (workspace.snsStatusMessage ?? "").length > 0;
    case "rds":
      return (workspace.rdsRegions?.length ?? 0) > 0 ||
        (workspace.rdsStatusMessage ?? "").length > 0;
    case "ecs":
      return (workspace.ecsRegions?.length ?? 0) > 0 ||
        (workspace.ecsStatusMessage ?? "").length > 0;
    case "eks":
      return (workspace.eksRegions?.length ?? 0) > 0 ||
        (workspace.eksStatusMessage ?? "").length > 0;
    case "cloudformation":
      return (workspace.cloudFormationRegions?.length ?? 0) > 0 ||
        (workspace.cloudFormationStatusMessage ?? "").length > 0;
    case "eventbridge":
      return (workspace.eventBridgeRegions?.length ?? 0) > 0 ||
        (workspace.eventBridgeStatusMessage ?? "").length > 0;
    case "route53":
      return (workspace.route53HostedZones?.length ?? 0) > 0 ||
        (workspace.route53StatusMessage ?? "").length > 0;
    case "elb":
      return (workspace.elbRegions?.length ?? 0) > 0 ||
        (workspace.elbStatusMessage ?? "").length > 0;
    case "apigateway":
      return (workspace.apiGatewayRegions?.length ?? 0) > 0 ||
        (workspace.apiGatewayStatusMessage ?? "").length > 0;
    case "secrets":
      return (workspace.secretsManagerRegions?.length ?? 0) > 0 ||
        (workspace.secretsManagerStatusMessage ?? "").length > 0;
    case "logs":
      return (workspace.logsRegions?.length ?? 0) > 0 ||
        (workspace.logsStatusMessage ?? "").length > 0;
    case "iam":
      return (workspace.iamRoles?.length ?? 0) > 0 ||
        (workspace.iamPolicies?.length ?? 0) > 0 ||
        (workspace.iamStatusMessage ?? "").length > 0;
    default:
      return false;
  }
}

File: backend/daemon/internal/models/aws_inventory.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Ali Shaikh

package models

// AwsInventorySlice is the scoped response returned by aws.inventory.get.
// Exactly one embedded payload in Payload is populated, as selected by Scope.
type AwsInventorySlice struct {
	ProviderID string              `json:"providerId"`
	Scope      string              `json:"scope"`
	Payload    AwsInventoryPayload `json:"payload"`
}

// AwsInventoryPayload is a closed set of flat AWS inventory payloads. Anonymous
// pointers preserve the existing WorkspaceSnapshot field names on the wire
// while ensuring unrelated scopes are omitted.
type AwsInventoryPayload struct {
	*AwsS3InventoryPayload
	*AwsEc2InventoryPayload
	*AwsLambdaInventoryPayload
	*AwsDynamoDBInventoryPayload
	*AwsSqsInventoryPayload
	*AwsSnsInventoryPayload
	*AwsRdsInventoryPayload
	*AwsEcsInventoryPayload
	*AwsEksInventoryPayload
	*AwsCloudFormationInventoryPayload
	*AwsEventBridgeInventoryPayload
	*AwsRoute53InventoryPayload
	*AwsElbInventoryPayload
	*AwsKmsInventoryPayload
	*AwsApiGatewayInventoryPayload
	*AwsSecretsManagerInventoryPayload
	*AwsLogsInventoryPayload
	*AwsIamInventoryPayload
}

type AwsS3InventoryPayload struct {
	SelectedS3BucketName string               `json:"selectedS3BucketName,omitempty"`
	SelectedS3ObjectKey  string               `json:"selectedS3ObjectKey,omitempty"`
	S3PrefixFilter       string               `json:"s3PrefixFilter,omitempty"`
	S3StatusMessage      string               `json:"s3StatusMessage,omitempty"`
	S3Buckets            []AwsS3Bucket        `json:"s3Buckets"`
	S3Objects            []AwsS3Object        `json:"s3Objects"`
	S3ObjectsNextToken   string               `json:"s3ObjectsNextToken,omitempty"`
	S3ObjectsHasMore     bool                 `json:"s3ObjectsHasMore,omitempty"`
	S3ObjectMetadata     []DetailField        `json:"s3ObjectMetadata"`
	S3ExportSnippets     []AwsS3ExportSnippet `json:"s3ExportSnippets"`
}

type AwsEc2InventoryPayload struct {
	SelectedEC2Region     string           `json:"selectedEc2Region,omitempty"`
	SelectedEC2InstanceID string           `json:"selectedEc2InstanceId,omitempty"`
	EC2StatusMessage      string           `json:"ec2StatusMessage,omitempty"`
	EC2Regions            []string         `json:"ec2Regions"`
	EC2Instances          []AwsEc2Instance `json:"ec2Instances"`
}

type AwsLambdaInventoryPayload struct {
	SelectedLambdaRegion       string              `json:"selectedLambdaRegion,omitempty"`
	SelectedLambdaFunctionName string              `json:"selectedLambdaFunctionName,omitempty"`
	LambdaStatusMessage        string              `json:"lambdaStatusMessage,omitempty"`
	LambdaRegions              []string            `json:"lambdaRegions"`
	LambdaFunctions            []AwsLambdaFunction `json:"lambdaFunctions"`
}

type AwsDynamoDBInventoryPayload struct {
	SelectedDynamoDBRegion    string             `json:"selectedDynamodbRegion,omitempty"`
	SelectedDynamoDBTableName string             `json:"selectedDynamodbTableName,omitempty"`
	DynamoDBStatusMessage     string             `json:"dynamodbStatusMessage,omitempty"`
	DynamoDBRegions           []string           `json:"dynamodbRegions"`
	DynamoDBTables            []AwsDynamoDBTable `json:"dynamodbTables"`
}

type AwsSqsInventoryPayload struct {
	SelectedSQSRegion   string        `json:"selectedSqsRegion,omitempty"`
	SelectedSQSQueueURL string        `json:"selectedSqsQueueUrl,omitempty"`
	SQSStatusMessage    string        `json:"sqsStatusMessage,omitempty"`
	SQSRegions          []string      `json:"sqsRegions"`
	SQSQueues           []AwsSqsQueue `json:"sqsQueues"`
}

type AwsSnsInventoryPayload struct {
	SelectedSNSRegion   string        `json:"selectedSnsRegion,omitempty"`
	SelectedSNSTopicArn string        `json:"selectedSnsTopicArn,omitempty"`
	SNSStatusMessage    string        `json:"snsStatusMessage,omitempty"`
	SNSRegions          []string      `json:"snsRegions"`
	SNSTopics           []AwsSnsTopic `json:"snsTopics"`
}

type AwsRdsInventoryPayload struct {
	SelectedRDSRegion     string           `json:"selectedRdsRegion,omitempty"`
	SelectedRDSInstanceID string           `json:"selectedRdsInstanceId,omitempty"`
	RDSStatusMessage      string           `json:"rdsStatusMessage,omitempty"`
	RDSRegions            []string         `json:"rdsRegions"`
	RDSInstances          []AwsRdsInstance `json:"rdsInstances"`
}

type AwsEcsInventoryPayload struct {
	SelectedECSRegion     string          `json:"selectedEcsRegion,omitempty"`
	SelectedECSClusterArn string          `json:"selectedEcsClusterArn,omitempty"`
	SelectedECSServiceArn string          `json:"selectedEcsServiceArn,omitempty"`
	SelectedECSTaskArn    string          `json:"selectedEcsTaskArn,omitempty"`
	ECSStatusMessage      string          `json:"ecsStatusMessage,omitempty"`
	ECSRegions            []string        `json:"ecsRegions"`
	ECSClusters           []AwsEcsCluster `json:"ecsClusters"`
	ECSServices           []AwsEcsService `json:"ecsServices"`
	ECSTasks              []AwsEcsTask    `json:"ecsTasks"`
}

type AwsEksInventoryPayload struct {
	SelectedEKSRegion      string            `json:"selectedEksRegion,omitempty"`
	SelectedEKSClusterName string            `json:"selectedEksClusterName,omitempty"`
	EKSStatusMessage       string            `json:"eksStatusMessage,omitempty"`
	EKSRegions             []string          `json:"eksRegions"`
	EKSClusters            []AwsEksCluster   `json:"eksClusters"`
	EKSNodeGroups          []AwsEksNodeGroup `json:"eksNodeGroups"`
}

type AwsCloudFormationInventoryPayload struct {
	SelectedCloudFormationRegion    string                        `json:"selectedCloudFormationRegion,omitempty"`
	SelectedCloudFormationStackName string                        `json:"selectedCloudFormationStackName,omitempty"`
	CloudFormationStatusMessage     string                        `json:"cloudFormationStatusMessage,omitempty"`
	CloudFormationRegions           []string                      `json:"cloudFormationRegions"`
	CloudFormationStacks            []AwsCloudFormationStack      `json:"cloudFormationStacks"`
	CloudFormationStackEvents       []AwsCloudFormationStackEvent `json:"cloudFormationStackEvents"`
}

type AwsEventBridgeInventoryPayload struct {
	SelectedEventBridgeRegion  string               `json:"selectedEventBridgeRegion,omitempty"`
	SelectedEventBridgeBusName string               `json:"selectedEventBridgeBusName,omitempty"`
	EventBridgeStatusMessage   string               `json:"eventBridgeStatusMessage,omitempty"`
	EventBridgeRegions         []string             `json:"eventBridgeRegions"`
	EventBridgeBuses           []AwsEventBridgeBus  `json:"eventBridgeBuses"`
	EventBridgeRules           []AwsEventBridgeRule `json:"eventBridgeRules"`
}

type AwsRoute53InventoryPayload struct {
	SelectedRoute53HostedZoneID string                        `json:"selectedRoute53HostedZoneId,omitempty"`
	Route53StatusMessage        string                        `json:"route53StatusMessage,omitempty"`
	Route53HostedZones          []AwsRoute53HostedZone        `json:"route53HostedZones"`
	Route53ResourceRecordSets   []AwsRoute53ResourceRecordSet `json:"route53ResourceRecordSets"`
}

type AwsElbInventoryPayload struct {
	SelectedElbRegion          string               `json:"selectedElbRegion,omitempty"`
	SelectedElbLoadBalancerArn string               `json:"selectedElbLoadBalancerArn,omitempty"`
	ElbStatusMessage           string               `json:"elbStatusMessage,omitempty"`
	ElbRegions                 []string             `json:"elbRegions"`
	ElbLoadBalancers           []AwsElbLoadBalancer `json:"elbLoadBalancers"`
	ElbTargetGroups            []AwsElbTargetGroup  `json:"elbTargetGroups"`
}

type AwsKmsInventoryPayload struct {
	SelectedKmsRegion string        `json:"selectedKmsRegion,omitempty"`
	SelectedKmsKeyId  string        `json:"selectedKmsKeyId,omitempty"`
	KmsStatusMessage  string        `json:"kmsStatusMessage,omitempty"`
	KmsRegions        []string      `json:"kmsRegions"`
	KmsKeys           []AwsKmsKey   `json:"kmsKeys"`
	KmsAliases        []AwsKmsAlias `json:"kmsAliases"`
}

type AwsApiGatewayInventoryPayload struct {
	SelectedApiGatewayRegion string               `json:"selectedApiGatewayRegion,omitempty"`
	SelectedApiGatewayApiKey string               `json:"selectedApiGatewayApiKey,omitempty"`
	ApiGatewayStatusMessage  string               `json:"apiGatewayStatusMessage,omitempty"`
	ApiGatewayRegions        []string             `json:"apiGatewayRegions"`
	ApiGatewayApis           []AwsApiGatewayApi   `json:"apiGatewayApis"`
	ApiGatewayStages         []AwsApiGatewayStage `json:"apiGatewayStages"`
}

type AwsSecretsManagerInventoryPayload struct {
	SelectedSecretsManagerRegion string                    `json:"selectedSecretsManagerRegion,omitempty"`
	SelectedSecretsManagerName   string                    `json:"selectedSecretsManagerName,omitempty"`
	SecretsManagerStatusMessage  string                    `json:"secretsManagerStatusMessage,omitempty"`
	SecretsManagerRegions        []string                  `json:"secretsManagerRegions"`
	SecretsManagerSecrets        []AwsSecretsManagerSecret `json:"secretsManagerSecrets"`
}

type AwsLogsInventoryPayload struct {
	SelectedLogsRegion   string        `json:"selectedLogsRegion,omitempty"`
	SelectedLogGroupName string        `json:"selectedLogGroupName,omitempty"`
	LogsStatusMessage    string        `json:"logsStatusMessage,omitempty"`
	LogsRegions          []string      `json:"logsRegions"`
	LogGroups            []AwsLogGroup `json:"logGroups"`
}

type AwsIamInventoryPayload struct {
	SelectedIAMRoleName string         `json:"selectedIamRoleName,omitempty"`
	IAMStatusMessage    string         `json:"iamStatusMessage,omitempty"`
	IAMRoles            []AwsIamRole   `json:"iamRoles"`
	IAMPolicies         []AwsIamPolicy `json:"iamPolicies"`
}
